src/algorithms/optimization/fidelity_initialize.jl

```diff
@@ -0,0 +1,145 @@
+"""
+    single_site_fidelity_initialize(
+        peps₀::InfinitePEPS, envspace, [bondspace = _maxspace(peps₀)];
+        noise_amp = 1.0e-2, kwargs...
+    )
+
+Generate a single-site unit cell PEPS from a (possibly) multi-site `peps₀` by maximizing
+the fidelity w.r.t. `peps₀`. Here, `envspace` determines the virtual environment spaces for
+CTMRG contractions. By default, the maximal bond space of `peps₀` is used for all virtual
+legs of the single-site PEPS.
+
+The single-site PEPS is intialized with Gaussian nosie and multiplied by `noise_amp`.
+The `kwargs...` are passed onto the [`approximate!`](@ref) call, refer to the docs
+for further details.
+"""
+function single_site_fidelity_initialize(
+        peps₀::InfinitePEPS, envspace, bondspace = _spacemax(peps₀);
+        noise_amp = 1.0e-1, kwargs...
+    )
+    @assert allequal(map(p -> space(p, 1), unitcell(peps₀))) "PEPS must have uniform physical spaces"
+
+    physspace = space(unitcell(peps₀)[1], 1)
+    peps_single = noise_amp * InfinitePEPS(randn, scalartype(peps₀), physspace, bondspace) # single-site unit cell with random noise
+
+    peps_uc = InfinitePEPS(fill(only(unitcell(peps_single)), size(peps₀))) # fill peps₀ unit cell with peps_singles
+    peps_single, env_single = approximate!(peps_uc, peps₀, envspace; kwargs...) # modifies peps_single in-place
+
+    return peps_single, env_single
+end
+
+# maximal virtual space over unit cell
+function _spacemax(peps::InfinitePEPS)
+    return reduce(supremum, map(p -> supremum(domain(p)[1], domain(p)[2]), unitcell(peps)))
+end
+
+"""
+    approximate!(
+        pepsdst::InfinitePEPS, pepssrc::InfinitePEPS, envspace;
+        maxiter = 5, tol = 1.0e-3, boundary_alg=(; verbosity=1)
+    )
+
+Approximate `pepssrc` with `pepsdst` by iteratively maximizing their fidelity where the
+contents of `pepssrc` are embedded into `pepsdst`. To contract the respective networks, the
+specified `envspace` is used on the environment bonds and kept fixed. The CTMRG contraction
+algorithm is specified via the `boundary_alg` `NamedTuple`.
+"""
+function MPSKit.approximate!(
+        pepsdst::InfinitePEPS, pepssrc::InfinitePEPS, envspace;
+        maxiter = 10, tol = 1.0e-3, boundary_alg = (; verbosity = 1)
+    )
+    @assert size(pepsdst) == size(pepssrc) "incompatible unit cell sizes"
+
+    # absorb src PEPS tensors into dst tensors in-place
+    for (pdst, psrc) in zip(unitcell(pepsdst), unitcell(pepssrc))
+        absorb!(pdst, psrc)
+    end
```

Comment:
I think it would be better to have this in `single_site_fidelity_initialize` instead of inside the actual `approximate!` implementation. Now we're really throwing away the initial guess, which is not really what we want if we start from a very good guess already for example.

Reply:
My concern wasn't so much the fact that it was in-place, but rather that doing the absorption step by default pretty much discards the initial guess. I would prefer to be able to actually use a really good initial guess for `pepsdst`.

So definitely agree with having a  non-mutating `approximate`, but even there much of the initial guess is thrown away by the `absorb` step. Similar to the MPSKit algorithms, it seems better to really start from the initial guess that was given as is, and keep the step that prepares a good initial guess for the routine calling `approximate`, i.e., `single_site_fidelity_initialize`.